storage/src/adb/current.rs

```diff
@@ -195,116 +207,164 @@ impl<
+                break;
+            }
+            let op = self.any.log.read(self.any.inactivity_floor_loc).await?;
+            let old_loc = self
+                .any
+                .move_op_if_active(op, self.any.inactivity_floor_loc)
+                .await?;
+            if let Some(old_loc) = old_loc {
+                let mut g_hasher = GHasher::new(&mut self.any.hasher, 9, &self.any.ops);
+                self.status.set_bit(&mut g_hasher, old_loc, false).await?;
+                self.status.append(&mut g_hasher, true).await?;
+            }
+            self.any.inactivity_floor_loc += 1;
+        }
+
+        self.any
+            .apply_op(Operation::Commit(self.any.inactivity_floor_loc))
+            .await?;
+        let mut g_hasher = GHasher::new(&mut self.any.hasher, 9, &self.any.ops);
+        self.status.append(&mut g_hasher, false).await?;
+
+        Ok(())
+    }
 
-        // Prune inactive elements from the any db.
-        self.any.prune_inactive().await?;
+    /// Commit any pending operations to the db, ensuring they are persisted to disk & recoverable
+    /// upon return from this function. Also raises the inactivity floor according to the schedule,
+    /// and prunes those operations below it.
+    pub async fn commit(&mut self) -> Result<(), Error> {
+        // Failure recovery relies on this specific order of these three disk-based operations:
+        //  (1) commit/sync the any db to disk (which raises the inactivity floor).
+        //  (2) prune the bitmap to the updated inactivity floor and write its state to disk.
+        //  (3) prune the any db of inactive operations.
+        self.commit_ops().await?; // (1)
 
-        // Failure recovery code relies on the bitmap getting written & pruned *after* ops have been
-        // committed & the adb::any pruned.
         self.status.prune_to_bit(self.any.inactivity_floor_loc);
         self.status
             .write_pruned(
                 self.context.with_label("bitmap"),
                 &self.bitmap_metadata_partition,
             )
-            .await?;
+            .await?; // (2)
+
+        // Prune inactive elements from the any db. We do this last, because bitmap recovery could
+        // require access to the hashes of these inactive nodes due to node grafting.
+        self.any.prune_inactive().await?; // (3)
 
         Ok(())
     }
 
-    /// Return the root hash of the db.
+    /// Return the root of the db.
     ///
     /// Current implementation just hashes the roots of the [Any] and [Bitmap] databases together.
-    pub fn root(&self, hasher: &mut H) -> H::Digest {
-        let any_root = self.any.root(hasher);
-        let bitmap_root = self.status.root(hasher);
+    pub async fn root(&self, hasher: &mut H) -> Result<H::Digest, Error> {
+        let mut b_hasher = Basic::new(hasher);
+        let any_root = self.any.root(&mut b_hasher);
+
+        let mut g_hasher = GHasher::new(hasher, 9, &self.any.ops);
```

Comment:
nit: `9` hardcoded again?

Reply:
yup a few other locations as well, all should be fixed now.